### PR TITLE
Auto-detect Helm repos + support Helm basic auth + fix bugs

### LINKED ISSUE #2260: Helm: auto-detect URLs
**Is your feature request related to a problem? Please describe.**

We currently require you to enter the Helm URL in the configuration for the URL to be recognised as Helm. This doesn't make it easy for either well known Helm repos.

**Describe the solution you'd like**

Auto-detect Helm repo URLs.

See #1865

### PR DESCRIPTION
**Pull request by alexec**
Closes #2260. See #2261.

Auto-detecting repo type:
```bash
$ argocd repo add https://kubernetes-charts.storage.googleapis.com --name stable
repository 'https://kubernetes-charts.storage.googleapis.com' added
$ argocd repo list
TYPE  NAME    REPO                                                     INSECURE  LFS    USER  STATUS      MESSAGE
git           git@github.intuit.com:acollins8/argocd-example-apps.git  false     false  -     Successful  
helm  stable  https://kubernetes-charts.storage.googleapis.com         false     false  -     Successful  
```

### CODE REVIEW

File: util/helm/repo/index.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+type entry struct {
+	Version string
+	Created time.Time
+}
+
+type index struct {
+	Entries map[string][]entry
+}
+
+func Index(url string) (*index, error) {
```
Comment (alexec) [inline util/helm/repo/index.go]: this is just a lift-and-shift

File: util/helm/repo/index.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	start := time.Now()
+
+	req, err := http.NewRequest("GET", url+"/index.yaml", nil)
+	if err != nil {
+		return nil, err
+	}
+	if username != "" {
+		// only basic supported
```
Comment (alexec) [inline util/helm/repo/index.go]: new auth code

File: util/helm/repo/repo_factory.go [hunk 3]
```diff
@@ -50,11 +50,6 @@ func NewRepo(url, name, username, password string, caData, certData, keyData []b
 		certData: certData,
 		keyData:  keyData,
 	}
-	err = r.Init()
```
Comment (alexec) [inline util/helm/repo/repo_factory.go]: double-init was happening, 2x the time it take to create a Helm repo

File: util/helm/repo/repo_factory_test.go [hunk 4]
```diff
@@ -6,17 +6,11 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestRepoFactory(t *testing.T) {
+func TestNewRepo(t *testing.T) {
 	t.Run("Unnamed", func(t *testing.T) {
 		_, err := NewRepo("http://0.0.0.0", "", "", "", nil, nil, nil)
 		assert.EqualError(t, err, "must name repo")
 	})
-
```
Comment (alexec) [inline util/helm/repo/repo_factory_test.go]: `NewRepo` is dumber now

File: util/helm/repo/index.go (hunk 1 above)
Comment (alexec) [reply util/helm/repo/index.go]: now I've added auth

**Comment by alexec** [comment]
`TestCustomToolWithGitCreds` maybe flaky. Re-running.

File: pkg/apis/application/v1alpha1/types.go [hunk 5]
```diff
@@ -1005,20 +1005,26 @@ func (repo *Repository) IsLFSEnabled() bool {
...
 	if source != nil {
-		m.Username = source.Username
-		m.Password = source.Password
-		m.SSHPrivateKey = source.SSHPrivateKey
-		m.InsecureIgnoreHostKey = source.InsecureIgnoreHostKey
-		m.Insecure = source.Insecure
-		m.EnableLFS = source.EnableLFS
-		m.TLSClientCertData = source.TLSClientCertData
-		m.TLSClientCertKey = source.TLSClientCertKey
+		if m.Username == "" {
```
Comment (alexec) [inline pkg/apis/application/v1alpha1/types.go]: tests

File: server/repository/repository.go [hunk 6]
```diff
@@ -256,7 +264,16 @@ func (s *Server) ValidateAccess(ctx context.Context, q *repositorypkg.RepoAccess
 		TLSClientCertKey:  q.TlsClientCertKey,
 		TLSClientCAData:   q.TlsClientCAData,
 	}
-	_, err := factory.NewFactory().NewRepo(repo, metrics.NopReporter)
+	creds, err := s.db.GetRepository(ctx, q.Repo)
```
Comment (alexec) [inline server/repository/repository.go]: bug in existing code

File: pkg/apis/application/v1alpha1/types.go [hunk 7]
```diff
@@ -1005,20 +1005,29 @@ func (repo *Repository) IsLFSEnabled() bool {
 	return repo.EnableLFS
 }
 
-func (m *Repository) HasCredentials() bool {
-	return m.Username != "" || m.Password != "" || m.SSHPrivateKey != "" || m.InsecureIgnoreHostKey
-}
-
 func (m *Repository) CopyCredentialsFrom(source *Repository) {
 	if source != nil {
-		m.Username = source.Username
```
Comment (alexec) [inline pkg/apis/application/v1alpha1/types.go]: this no longer overwrites things you don't want overwritten

File: server/repository/repository.go [hunk 8]
```diff
@@ -184,13 +186,19 @@ func (s *Server) Create(ctx context.Context, q *repositorypkg.RepoCreateRequest)
 			return nil, err
 		}
 		repo.CopyCredentialsFrom(creds)
-		_, err = factory.NewFactory().NewRepo(q.Repo, metrics.NopReporter)
```
Comment (alexec) [inline server/repository/repository.go]: fixed bug

**Review by alexmt — APPROVED** [review]

**Pull request merged**